plugins/producer_plugin/include/eosio/producer_plugin/snapshot_scheduler.hpp

```diff
@@ -0,0 +1,186 @@
+
+class snapshot_scheduler {
+private:
+   struct by_snapshot_id;
+   struct by_snapshot_value;
+   struct as_vector;
+
+   using snapshot_requests = bmi::multi_index_container<
+         producer_plugin::snapshot_schedule_information,
+         indexed_by<
+               bmi::hashed_unique<tag<by_snapshot_id>, BOOST_MULTI_INDEX_MEMBER(producer_plugin::snapshot_request_id_information, uint32_t, snapshot_request_id)>,
+               bmi::random_access<tag<as_vector>>,
+               bmi::ordered_unique<tag<by_snapshot_value>,
+                                   composite_key<producer_plugin::snapshot_schedule_information,
+                                                 BOOST_MULTI_INDEX_MEMBER(producer_plugin::snapshot_request_information, uint32_t, block_spacing),
+                                                 BOOST_MULTI_INDEX_MEMBER(producer_plugin::snapshot_request_information, uint32_t, start_block_num),
+                                                 BOOST_MULTI_INDEX_MEMBER(producer_plugin::snapshot_request_information, uint32_t, end_block_num)>>>>;
+   snapshot_requests _snapshot_requests;
+   snapshot_db_json _snapshot_db;
+   uint32_t _snapshot_id   = 0;
+   uint32_t _inflight_sid  = 0;
+   std::function<void(producer_plugin::next_function<producer_plugin::snapshot_information>)> _create_snapshot;
+
+   void x_serialize() {
+      auto& vec = _snapshot_requests.get<as_vector>();
+      std::vector<producer_plugin::snapshot_schedule_information> sr(vec.begin(), vec.end());
+      _snapshot_db << sr;   
+   }
+
+public:
+   snapshot_scheduler() = default;
+
+   // snapshot_scheduler_listener
+   void on_start_block(uint32_t height) {
+      bool serialize_needed = false;
+
+      for(const auto& req: _snapshot_requests.get<0>()) {
+         // assume "asap" for snapshot with missed/zero start, it can have spacing
+         if(!req.start_block_num) {
+            // update start_block_num with current height only if this is recurring
+            // if non recurring, will be executed and unscheduled
+            if (req.block_spacing && height) {
+               auto& snapshot_by_id = _snapshot_requests.get<by_snapshot_id>();
+               auto it = snapshot_by_id.find(req.snapshot_request_id);
+               _snapshot_requests.modify(it, [&height](auto& p) { p.start_block_num = height; });
+               serialize_needed = true;              
+            }
+            execute_snapshot(req.snapshot_request_id);
+         }
+         // execute snapshot, -1 since its called from start block
+         else if(!req.block_spacing || (!((height - req.start_block_num - 1) % req.block_spacing))) {
+            execute_snapshot(req.snapshot_request_id);
+         }       
+
+         // cleanup - remove expired (or invalid) request
+         if(!req.start_block_num || !req.block_spacing || (req.end_block_num > 0 && height >= req.end_block_num)) {
+            unschedule_snapshot(req.snapshot_request_id);
```

Comment:
suggestion: add `serialize_needed = true;` here and remove it from `unschedule_snapshot()`

Reply:
unschedule_snapshot is also called directly from producer plugin that  expects serialization to happen as well

Reply:
OK, I missed that. So we could leave it as is, or add a parameter `bool serialize` to `unschedule_snapshot`. Up to you!